src/server/req_runjob.c

```diff
@@ -584,37 +584,52 @@ req_runjob(struct batch_request *preq)
 		return;
 
 	} else if (jt == IS_ARRAY_Single) {
-		attribute sub_runcount = {0};
-		attribute sub_run_version = {0};
+		attribute sub_runcount;
+		attribute sub_run_version;
 		attribute sub_prev_res;
+
+		clear_attr(&sub_runcount, &job_attr_def[JOB_ATR_runcount]);
+		clear_attr(&sub_run_version, &job_attr_def[JOB_ATR_run_version]);
 		clear_attr(&sub_prev_res, &job_attr_def[JOB_ATR_resource]);
 
 		pjobsub = get_subjob_and_state(parent, get_index_from_jid(jid), NULL, NULL);
 		if (pjobsub != NULL) {
-			sub_runcount = pjobsub->ji_wattr[JOB_ATR_runcount];
-			sub_run_version = pjobsub->ji_wattr[JOB_ATR_run_version];
+			if (is_jattr_set(pjobsub, JOB_ATR_runcount))
+				set_attr_with_attr(&job_attr_def[JOB_ATR_runcount], &sub_runcount, get_jattr(pjobsub, JOB_ATR_runcount), SET);
+			if (is_jattr_set(pjobsub, JOB_ATR_run_version))
+				set_attr_with_attr(&job_attr_def[JOB_ATR_run_version], &sub_run_version, get_jattr(pjobsub, JOB_ATR_run_version), SET);
```

Comment:
this also seems like it's changing logic. We were just setting pointers to the existing attributes before, now we are making a copy of it and freeing it in various places later. The code looks a lot messier now. Why did you have to do this?

Reply:
Just answering question:
There was bug in earlier code as it was saving pointer to existing area (aka from old pjobsub->ji_attr) but after then it calls job_purge(psubjob) which frees those saved pointers

Now in new code we make copy of needed attrs then use it later so even after job_purge() will have access to old values of saved attrs.

FYI: This is job rerun scenario.

Reply:
Thanks for clarifying Hiren. It sounds like a memory error so I think that it would be better to fix it independently (with valgrind traces etc.) and keep this PR limited to just refactoring.